service/matching/db_task_manager.go

```diff
@@ -0,0 +1,258 @@
+	logger log.Logger,
+	dispatchTaskFn func(*internalTask) error,
+	finishTaskFn func(*persistencespb.AllocatedTaskInfo, error),
+) (*dbTaskManager, error) {
+	taskOwnership := newDBTaskOwnership(
+		taskQueueKey,
+		taskQueueKind,
+		taskIDRangeSize,
+		store,
+		logger,
+	)
+	if err := taskOwnership.takeTaskQueueOwnership(); err != nil {
+		return nil, err
+	}
+
+	return &dbTaskManager{
+		status:        common.DaemonStatusInitialized,
+		store:         store,
+		taskOwnership: taskOwnership,
+		taskFlusher: newDBTaskFlusher(
+			taskQueueKey,
+			taskOwnership,
+			logger,
+		),
+		taskTracker: newDBTaskTracker(
+			taskQueueKey,
+			store,
+			taskOwnership.getAckedTaskID(),
+			logger,
+		),
+		dispatchTaskFn: dispatchTaskFn,
+		finishTaskFn:   finishTaskFn,
+		logger:         logger,
+
+		shutdownChan: make(chan struct{}),
+		dispatchChan: make(chan struct{}, 1),
+	}, nil
+}
+
+func (d *dbTaskManager) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	d.SignalDispatch()
+	go d.readerEventLoop()
+	go d.writerEventLoop()
+}
+
+func (d *dbTaskManager) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(d.shutdownChan)
+}
+
+func (d *dbTaskManager) SignalDispatch() {
+	select {
+	case d.dispatchChan <- struct{}{}:
+	default: // channel already has an event, don't block
+	}
+}
+
+func (d *dbTaskManager) isStopped() bool {
+	return atomic.LoadInt32(&d.status) == common.DaemonStatusStopped
+}
+
+func (d *dbTaskManager) writerEventLoop() {
+	updateQueueTicker := time.NewTicker(dbTaskUpdateQueueInterval)
+	defer updateQueueTicker.Stop()
+	// TODO we should impl a more efficient method to
+	//  buffer & wait for max duration
+	//  right now simply just flush every dbTaskFlushInterval
+	flushTicker := time.NewTicker(dbTaskFlushInterval)
+	defer flushTicker.Stop()
+
+	for {
+		if d.isStopped() {
+			return
+		}
+
+		select {
+		case <-d.shutdownChan:
+			return
+		case <-d.taskOwnership.getShutdownChan():
+			d.Stop()
+
+		case <-updateQueueTicker.C:
+			d.persistTaskQueue()
+		case <-flushTicker.C:
+			d.taskFlusher.flushTasks()
+			d.SignalDispatch()
+		case <-d.taskFlusher.notifyFlushChan():
+			d.taskFlusher.flushTasks()
+			d.SignalDispatch()
+		}
+	}
+}
+
+func (d *dbTaskManager) readerEventLoop() {
+	updateAckTicker := time.NewTicker(dbTaskUpdateAckInterval)
+	defer updateAckTicker.Stop()
+
+	for {
+		if d.isStopped() {
+			return
+		}
+
+		select {
+		case <-d.shutdownChan:
+			return
+		case <-d.taskOwnership.getShutdownChan():
+			d.Stop()
+
+		case <-updateAckTicker.C:
+			d.updateAckTaskID()
+		case <-d.dispatchChan:
+			d.readDispatchTask()
+		}
+	}
+}
+
+func (d *dbTaskManager) writeAppendTask(
+	task *persistencespb.TaskInfo,
+) future.Future {
+	return d.taskFlusher.appendTask(task)
+}
+
+func (d *dbTaskManager) readDispatchTask() {
```

Comment:
rename to dispatchTasks

Reply:
this function actually read then dispatch a task, not simply doing in mem dispatch

Reply:
maybe readAndDispatchTasks?